### PR TITLE
feature/issue-258 add bernoulli_logit_rng

### PR DESCRIPTION
**Pull request by thelseraphim**
#### Summary:

Adding Logit Parametrization for Bernoulli random number generators, resolving https://github.com/stan-dev/math/issues/258

#### Intended Effect:

Adds the ability to generate bernoulli random numbers in a different parametrization.

#### How to Verify:

Added a unit test at:

`./runTests.py test/unit/math/prim/scal/prob/bernoulli_test.cpp`

#### Side Effects:

None, function is not called. Will follow this with a pull request adding the function to stan.

#### Documentation:

Doxygen

#### Reviewer Suggestions: 

This is my first pull request on stan, so scrutiny is appreciated.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses: 
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by bob-carpenter** [comment]
Does "WIP" mean "work in progress"?   Does that mean it's not ready to review yet or did you want feedback on something?

We need a declared copyright holder in the pull request, which for you will be Columbia University (or the trustees thereof if you want to be formal).

**Comment by thelseraphim** [comment]
I'd appreciate review, I just wanted to indicate that it's probably not ready to merge because I haven't documented it.

File: test/unit/math/prim/scal/prob/bernoulli_test.cpp [hunk 1]
```diff
@@ -40,3 +42,33 @@ TEST(ProbDistributionsBernoulli, chiSquareGoodnessFitTest) {
...
+  boost::random::mt19937 rng;
+  int N = 10000;
+  boost::math::chi_squared mydist(1);
+ 
+  double parameter = -0.5;
+  double prob = stan::math::inv_logit(-0.5);
+  int bin[2] = {0, 0};
+  double expect [2] = {N * (1 - prob), N * prob};
+
+  int count = 0;
```
Comment (bob-carpenter) [inline test/unit/math/prim/scal/prob/bernoulli_test.cpp]: I'd recommend this:

```
int bin[2] = { 0, 0};
for (int count = 0; count < N; ++count)
  ++bin[bernoulli_logit_rng(parameter, rng)];
```

The for loop synchs up the initialization, condition and increment to make it easier to read.  And you can just use the value of the rng directly to define the bin here.

File: test/unit/math/prim/scal/prob/bernoulli_test.cpp [hunk 2]
```diff
@@ -40,3 +42,33 @@ TEST(ProbDistributionsBernoulli, chiSquareGoodnessFitTest) {
...
+  while (count < N) {
+    int a = stan::math::bernoulli_logit_rng(parameter,rng);
+    if(a == 1)
+      ++bin[1];
+    else
+      ++bin[0];
+    count++;
+   }
+
+  double chi = 0;
```
Comment (bob-carpenter) [inline test/unit/math/prim/scal/prob/bernoulli_test.cpp]: Not necessary, but it'd be great if you could convert to `std::vector` and pull this out into a utility test function

```
void chi_square_test(const std::vector<int>& expect, const std::vector<int>& bin);
```
that would declare the `mydist` variable internally and run the `EXPECT_TRUE`.  Then we could have another issue to clean up all the other rngs which probably do it this way.  We've created almost all of our functions that operate on sequences using `std::vector` rather than using arrays or pointers mainly for memory encapsulation.  It makes the second init harder and the first one easier.

File: stan/math/prim/scal/prob/bernoulli_logit_rng.hpp [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+#include <stan/math/prim/scal/fun/constants.hpp>
+#include <stan/math/prim/scal/fun/inv_logit.hpp>
+#include <stan/math/prim/scal/fun/log1m.hpp>
+#include <stan/math/prim/scal/fun/value_of.hpp>
+#include <stan/math/prim/scal/meta/include_summand.hpp>
+
+namespace stan {
+  namespace math {
+
+    template <class RNG>
```
Comment (bob-carpenter) [inline stan/math/prim/scal/prob/bernoulli_logit_rng.hpp]: needs doc

File: stan/math/prim/scal/prob/bernoulli_logit_rng.hpp [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    inline int
+    bernoulli_logit_rng(const double t,
```
Comment (bob-carpenter) [inline stan/math/prim/scal/prob/bernoulli_logit_rng.hpp]: Don't need const for primitive types as it doesn't change the external API.  It is usually good style in the body of the function to treat the arguments as constant, but that doesn't need to be exposed to the outside world.

**Comment by bob-carpenter** [comment]
I edited to move Columbia Uni up.  The form's too busy right now, but we wanted to make it clear to external pull requesters what was going on.

**Comment by thelseraphim** [comment]
Ok, general comment: none of the files in `stan/math/prim/scal/prob/*_rng.hpp` are doc'ed and all of them const primitive variables. `git blame` reveals that most of these files are very old. I'd like to just add this thing the old way, for consistency and make a new ticket for bringing them all up to our current coding style.

**Comment by thelseraphim** [comment]
Made all the mentioned changes except for the const thing, where I certainly don't mind removing it but prefer to keep it consistent in this PR and then clean them all up at once in a followup. If this is a mistake it's one that we make in every file.

**Comment by bob-carpenter** [comment]
OK, we can try to go through and do a const sweep in all of the code.  Indeed most of these are very old.  And we also need to go through and do a doc and testing sweep, because not everything's documented or tested well.  We're trying to get new code in the way we want it to the extent possible.  We definitely want doc and tests for everything new getting added and as much retrofitted doc and tests as possible.

File: test/unit/math/prim/scal/prob/bernoulli_test.cpp [hunk 5]
```diff
@@ -6,37 +6,68 @@
...
                std::domain_error);
+  EXPECT_THROW(stan::math::bernoulli_logit_rng(stan::math::positive_infinity(),rng),
+               std::domain_error);
+}
+
+/*
+ * Uses a chi-squared test to check that counts are reasonably consistent
+ * with draws from a Bernoulli distribution..
+ */
+void assert_bernoulli(const std::vector<int>& counts, const std::vector<double>& expected_counts, double tolerance) {
```
Comment (bob-carpenter) [inline test/unit/math/prim/scal/prob/bernoulli_test.cpp]: It's OK set to 1 in this test, which is the number of bins (2) minus one.

File: stan/math/prim/scal/prob/bernoulli_logit_rng.hpp [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+     * often more convenient logit-parametrization.
+     * 
+     * @param t logit-transformed probability parameter.
+     * @param rng pseudorandom number generator.
+     * @tparam RNG Random number generator type.
+     * @return Bernoulli(logit^{-1}(t)) generated random number, either 0 or 1.
+     */
+    template <class RNG>
+    inline int
+    bernoulli_logit_rng(const double t,
```
Comment (syclik) [inline stan/math/prim/scal/prob/bernoulli_logit_rng.hpp]: 1. remove `const`
2. don't know what happened with spacing, but move `RNG& rng` to the end of line 30. (Or make line 31 line up properly with line 30.

File: test/unit/math/prim/scal/prob/bernoulli_test.cpp [hunk 7]
```diff
@@ -6,37 +6,68 @@
 TEST(ProbDistributionsBernoulli, error_check) {
   boost::random::mt19937 rng;
   EXPECT_NO_THROW(stan::math::bernoulli_rng(0.6,rng));
+  EXPECT_NO_THROW(stan::math::bernoulli_logit_rng(-3.5,rng));
```
Comment (syclik) [inline test/unit/math/prim/scal/prob/bernoulli_test.cpp]: Revert the changes to `bernoulli_test.cpp` and this should all be in `bernoulli_logit_test.cpp`.

**Comment by syclik** [comment]
I'd prefer to fix the `const` argument with this pull request. There's no reason we shouldn't do things right going forward. By doing it correctly here, it won't change what we need to do with the sweep anyway.

File: stan/math/prim/scal/prob/bernoulli_logit_rng.hpp (hunk 6 above)
Comment (bob-carpenter) [inline stan/math/prim/scal/prob/bernoulli_logit_rng.hpp]: If you wouldn't mind, `tparam` doc should go first.  I'm OK with leaving the `const` if you really want to and fixing them all later.  Neither the doc orders or use of `in` and `out` to mark arguments that must be specified as input and/or are changed on output.  A lot of our early code doesn't have any `const` where there should be. I've been adding doc and fixing doc as we refactor and add code, but we need a big const correctness refactor for all of our generated model code.

File: stan/math/prim/scal/prob/bernoulli_logit_rng.hpp [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+                  RNG& rng) {
+      using boost::variate_generator;
+      using boost::bernoulli_distribution;
+      using stan::math::inv_logit;
+
+      static const char* function("bernoulli_logit_rng");
+
+      check_finite(function, "Logit transformed probability parameter", t);
```
Comment (syclik) [inline stan/math/prim/scal/prob/bernoulli_logit_rng.hpp]: Remove `function` and just use `"bernoulli_logit_rng"` directly.

Comment (syclik) [inline stan/math/prim/scal/prob/bernoulli_logit_rng.hpp]: didn't mean to unapprove changes. (I'm still working through GitHub's new interface)

File: test/unit/math/prim/scal/prob/bernoulli_test.cpp (hunk 5 above)
Comment (thelseraphim) [inline test/unit/math/prim/scal/prob/bernoulli_test.cpp]: Happy to make a general chi square test, where in the tree do you think it should live?

**Comment by thelseraphim** [comment]
Everything is fixed except for the absolutely silly repetition and narrow scope of that chi-square test. Happy to rewrite that more generally and put it in a place for other tests to use it if anyone has a suggestion for where to put the utility function. Otherwise should be ready to merge.

**Comment by bob-carpenter** [comment]
Let's pull the chi-square test out later for all the RNG tests.  I created the issue #412 for it.

**Comment by bob-carpenter** [comment]
I think @syclik still needs to approve the changes he requested.  I can merge without him doing that if he's not around when the tests pass.  Thanks for making all the fiddly little formatting and style changes; hopefully this'll be much easier for the next pull request.

**Comment by syclik** [comment]
Approved! Thanks @thelseraphim.

**Comment by syclik** [comment]
This pull request fixes #258.
